### ISSUE TITLE
TS2339 triggering inside if only with dynamic comparison

### ISSUE THREAD

**Issue by tonysamperi** [Microsoft/TypeScript#25087]
**TypeScript Version:**  3.0.0-dev.201xxxxx
Confirmed also in `next` version

**Search Terms:** 
TS2339, Property does not exist on type, Property

**Code**

```ts
            someFuncton (needle: User | string | null): any[]=> {
                let toReturn = [];
                // Case string
                if (typeof needle === typeof whitespace) {
                    toReturn = needle.length > 2 ? this._filterUsers(needle) : [];
                }
                // Case User
                else if (!!needle) {
                    return _.filter(this.usersList, needle)
                }
                // Case NULL
                else {
                       //Do something
                }
                return toReturn;
            })
```

**Expected behavior:**
No error is triggered

**Actual behavior:**
A TS2339 error is triggered: 
* error TS2339: Property 'length' does not exist on type 'string | User'.
* error TS2345: Argument of type 'string | User' is not assignable to parameter of type 'string'.

The second error is due to the fact that the _filterUsers function accepts `User` as argument type.

**Playground Link:**  
http://www.typescriptlang.org/play/#src=export%20class%20User%20%7B%0D%0A%20%20%20%20firstName%3A%20string%3B%0D%0A%20%20%20%20lastName%3A%20string%3B%0D%0A%20%20%20%20userId%3A%20string%3B%0D%0A%20%20%20%20division%3A%20string%3B%0D%0A%20%20%20%20businessUnit%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Aconst%20whitespace%20%3D%20%22%20%22%3B%0D%0A%0D%0Aexport%20function%20myCallback(needle%3A%20User%20%7C%20string%20%7C%20null)%20%7B%0D%0A%20%20%20%20let%20toReturn%20%3D%20%5B%5D%3B%0D%0A%20%20%20%20%2F%2F%20Case%20string%0D%0A%20%20%20%20if%20(typeof%20needle%20%3D%3D%3D%20typeof%20whitespace)%20%7B%0D%0A%20%20%20%20%20%20%20%20toReturn%20%3D%20needle.length%20%3E%202%20%3F%20%5B%22A%22%2C%20%22B%22%2C%20%22C%22%5D%20%3A%20%5B%5D%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%2F%2F%20Case%20User%0D%0A%20%20%20%20else%20if%20(!!needle)%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20Do%20something%20else%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20%2F%2F%20Case%20NULL%0D%0A%20%20%20%20else%20%7B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20Do%20awesome%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20return%20toReturn%3B%0D%0A%7D%3B

**Related Issues:** 
Not precisely like this

**Comment by tonysamperi** [comment]
Note: the problem solves if I use `typeof needle === "string"` or by adding a cast expression inside the if statement: `<string>needle.length`
